File: verify/page_parity.js
/* 주문 페이지(dist/spec.js) vs engine/strategy.py 대조.
 *
 *   python verify/page_parity.py && node verify/page_parity.js
 *
 * 두 구현이 같은 시나리오에서 같은 주문을 내는지 확인한다.
 * 페이지를 고칠 때마다 이 검사를 돌려 문서 스펙에서 떨어져 나가는 것을 막는다.
 */
'use strict';
const fs = require('fs'), path = require('path'), vm = require('vm');

const HERE = __dirname;
const DIST = path.join(HERE, '..', 'soxl-daily-orders-source', 'dist');
const EXP = path.join(HERE, '_parity_expected.json');

if (!fs.existsSync(EXP)) {
  console.error('기대값이 없습니다. 먼저 실행:  python verify/page_parity.py');
  process.exit(1);
}

// dist/data.js 와 dist/spec.js 를 브라우저처럼 같은 전역에 올린다
const sandbox = { window: {}, console };
sandbox.globalThis = sandbox;
vm.createContext(sandbox);
vm.runInContext(fs.readFileSync(path.join(DIST, 'data.js'), 'utf8'), sandbox);
vm.runInContext(fs.readFileSync(path.join(DIST, 'spec.js'), 'utf8'), sandbox);
const SEED = sandbox.window.SEED, SPEC = sandbox.window.SPEC;

const exp = JSON.parse(fs.readFileSync(EXP, 'utf8'));
const EPS = 1e-6;
let fails = 0, checks = 0;

function near(a, b) {
  if (a == null && b == null) return true;
  if (a == null || b == null) return false;
  return Math.abs(a - b) <= EPS * Math.max(1, Math.abs(a), Math.abs(b));
}
function cmp(label, got, want) {
  checks++;
  const ok = typeof want === 'number' ? near(got, want) : got === want;
  if (!ok) { fails++; console.log(`      [불일치] ${label}: JS ${got}  vs  PY ${want}`); }
  return ok;
}

// 1) 파라미터 대조
console.log('='.repeat(84));
console.log('1. 파라미터 대조  spec.js  vs  strategy.py');
console.log('='.repeat(84));
const P = SPEC.P, pp = exp.params;
const pmap = {
  W_BASE: P.W_BASE, W_STRONG: P.W_STRONG, BO_K: P.BO_K, BOS_K: P.BOS_K,
  BOS_RSI: P.BOS_RSI, BOS_MAX_W: P.BOS_MAX_W, G1_DROP: P.G1_DROP, EPS: P.EPS,
  MAX_DAYS: P.MAX_DAYS, TP_TOP: P.TP.TOP, TP_BOTTOM: P.TP.BOTTOM,
};
Object.keys(pmap).forEach(k => cmp(k, pmap[k], pp[k]));
pp.W_TOP.forEach((v, i) => cmp(`W_TOP[${i}]`, P.W_TOP[i], v));
pp.W_BOTTOM.forEach((v, i) => cmp(`W_BOTTOM[${i}]`, P.W_BOTTOM[i], v));
console.log(`  파라미터 ${checks}개 대조  ${fails ? fails + '건 불일치' : '전부 일치'}`);

// 2) 데이터 기준일
console.log('\n' + '='.repeat(84));
console.log('2. 데이터 기준일');
console.log('='.repeat(84));
cmp('asOf', SEED.soxl[SEED.soxl.length - 1][0], exp.asOf);
cmp('rsi', SEED.qqq.rsi, exp.rsi);
console.log(`  SOXL ${SEED.soxl.length}봉 / SOXS ${SEED.soxs.length}봉  기준일 ${SEED.qqq.date}  RSI ${SEED.qqq.rsi}`);

// 3) 시나리오별 주문 대조
console.log('\n' + '='.repeat(84));
console.log('3. 시나리오별 주문 대조');
console.log('='.repeat(84));
exp.cases.forEach((c, n) => {
  const before = fails;
  const got = SPEC.computeOrders({
    soxl: SEED.soxl, soxs: SEED.soxs,
    rsi: c.input.regime === 'BOTTOM' ? 40 : 58.1151,   // 레짐만 맞추면 된다
    cash: c.input.cash,
    boHold: { ticker: c.input.boQty > 0 ? c.input.boTick : null, qty: c.input.boQty },
    lots: c.input.lots, lastRung: c.input.lastRung,
  });
  const w = c.expect;
  cmp('총자산', got.total, w.total);
  cmp('목표비중', got.targetW, w.targetW);
  cmp('정배열', got.strong, w.strong);
  cmp('돌파현금', got.boCash, w.boCash);
  cmp('그리드현금', got.gridCash, w.gridCash);

  if (w.breakout) {
    cmp('돌파.종목', got.breakout && got.breakout.ticker, w.breakout.ticker);
    cmp('돌파.감시가', got.breakout && got.breakout.watch, w.breakout.watch);
    cmp('돌파.지정가', got.breakout && got.breakout.limit, w.breakout.limit);
    cmp('돌파.투입한도', got.breakout && got.breakout.budget, w.breakout.budget);
    cmp('돌파.수량', got.breakout && got.breakout.qty, w.breakout.qty);
  } else cmp('돌파', got.breakout, null);

  if (w.gridEntry) {
    cmp('진입.칸', got.gridEntry && got.gridEntry.rung, w.gridEntry.rung);
    cmp('진입.비중', got.gridEntry && got.gridEntry.weight, w.gridEntry.weight);
    cmp('진입.지정가', got.gridEntry && got.gridEntry.limit, w.gridEntry.limit);
    cmp('진입.금액', got.gridEntry && got.gridEntry.amount, w.gridEntry.amount);
    cmp('진입.수량', got.gridEntry && got.gridEntry.qty, w.gridEntry.qty);
  } else cmp('진입', got.gridEntry, null);

  cmp('청산 건수', got.gridExits.length, w.gridExits.length);
  w.gridExits.forEach((e, i) => {
    const g = got.gridExits[i] || {};
    cmp(`청산[${i}].칸`, g.no, e.no);
    cmp(`청산[${i}].지정가`, g.orderLimit, e.limit);
    cmp(`청산[${i}].경과`, g.elapsed, e.elapsed);
    cmp(`청산[${i}].강제`, g.forced, e.forced);
  });

  const ok = fails === before;
  console.log(`  ${ok ? '[일치]' : '[불일치]'}  ${n + 1}. ${c.name}`);
});

console.log('\n' + '='.repeat(84));
console.log(`대조 ${checks}건 중 불일치 ${fails}건 — ${fails ? '실패' : '주문 페이지와 파이썬 엔진이 동일'}`);
process.exit(fails ? 1 : 0);

File: verify/page_parity.py
# -*- coding: utf-8 -*-
"""주문 페이지(dist/spec.js) 대조용 기대값 생성.

engine/strategy.py 의 next_orders() 로 여러 시나리오의 주문을 계산해
verify/_parity_expected.json 으로 내보낸다. 그 다음 page_parity.js 가
같은 시나리오를 spec.js 로 계산해 두 값을 대조한다.

  python verify/page_parity.py && node verify/page_parity.js
"""
import os, sys, json
from datetime import date

H = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, os.path.join(H, '..', 'engine'))
import strategy as S
from strategy import load, next_orders

B = load()

# 페이지에 실린 시드와 같은 구간만 쓴다 (dist/data.js = 최근 260봉)
KEEP = 260
BW = B[-KEEP:]


def bar_index(d):
    for i, b in enumerate(BW):
        if b['d'].isoformat() == d:
            return i
    raise KeyError(d)


def lots(spec):
    """spec: [(칸번호, 매수가, 수량, 'YYYY-MM-DD', 레짐), ...]"""
    return [{'no': n, 'px': px, 'q': q, 'day_index': bar_index(d), 'rg': rg}
            for n, px, q, d, rg in spec]


DATES = [b['d'].isoformat() for b in BW]

SCENARIOS = [
    dict(name='정배열 아님 · 보유 없음 · TOP',
         cash=100000.0, bo_shares=0, bo_tick='SOXL', regime='TOP',
         lots=[], last_rung=0),
    dict(name='보유 없음 · BOTTOM 레짐',
         cash=100000.0, bo_shares=0, bo_tick='SOXL', regime='BOTTOM',
         lots=[], last_rung=0),
    dict(name='그리드 3칸 보유 · TOP',
         cash=42000.0, bo_shares=0, bo_tick='SOXL', regime='TOP',
         lots=[(1, 130.0, 90, DATES[-6], 'TOP'),
               (2, 126.5, 70, DATES[-4], 'TOP'),
               (3, 122.0, 55, DATES[-2], 'TOP')], last_rung=3),
    dict(name='중간 칸 청산 후 진행 (1,2 보유 · 마지막 매수 4번)',
         cash=30000.0, bo_shares=0, bo_tick='SOXL', regime='BOTTOM',
         lots=[(1, 130.0, 90, DATES[-9], 'TOP'),
               (2, 126.5, 70, DATES[-8], 'BOTTOM')], last_rung=4),
    dict(name='강제청산 임박 (7거래일째)',
         cash=15000.0, bo_shares=0, bo_tick='SOXL', regime='TOP',
         lots=[(1, 140.0, 60, DATES[-7], 'TOP'),
               (2, 133.0, 50, DATES[-3], 'BOTTOM')], last_rung=2),
    dict(name='전일 SOXL 돌파 보유 · 시가 청산 예정',
         cash=20000.0, bo_shares=150, bo_tick='SOXL', regime='TOP',
         lots=[(1, 128.0, 80, DATES[-5], 'TOP')], last_rung=1),
    dict(name='7칸 모두 보유 · 신규 진입 없음',
         cash=2000.0, bo_shares=0, bo_tick='SOXL', regime='TOP',
         lots=[(i, 140.0 - i * 3, 30, DATES[-8 + i], 'TOP') for i in range(1, 8)],
         last_rung=7),
    dict(name='현금 부족 · 목표금액이 현금에 구속',
         cash=1200.0, bo_shares=0, bo_tick='SOXL', regime='TOP',
         lots=[(1, 130.0, 200, DATES[-6], 'TOP')], last_rung=1),
]

out = []
for sc in SCENARIOS:
    # 페이지는 총 가용현금 하나만 받는다 -> 전액을 그리드 현금으로 넣고 재분배시킨다
    o = next_orders(BW, 0.0, sc['cash'], sc['bo_shares'], lots(sc['lots']),
                    regime=sc['regime'], bo_tick=sc['bo_tick'],
                    last_rung=sc['last_rung'])
    bo = o['돌파']; ge = o['그리드진입']
    out.append({
        'name': sc['name'],
        'input': {
            'cash': sc['cash'], 'boQty': sc['bo_shares'], 'boTick': sc['bo_tick'],
            'regime': sc['regime'], 'lastRung': sc['last_rung'],
            'lots': [{'no': n, 'px': px, 'qty': q, 'date': d, 'regime': rg}
                     for n, px, q, d, rg in sc['lots']],
        },
        'expect': {
            'total': o['총자산'], 'targetW': o['목표비중'], 'strong': o['정배열'],
            'boCash': o['리밸런싱']['돌파현금'], 'gridCash': o['리밸런싱']['그리드현금'],
            'breakout': None if not bo else {
                'ticker': bo['종목'], 'watch': bo['감시가'], 'limit': bo['지정가'],
                'budget': bo['투입한도'], 'qty': bo['수량']},
            'gridEntry': None if not ge else {
                'rung': ge['칸'], 'weight': ge['비중'], 'limit': ge['지정가'],
                'amount': ge['금액'], 'qty': ge['수량']},
            'gridExits': [{'no': x['칸'], 'limit': x['지정가'], 'elapsed': x['경과거래일'],
                           'forced': x['주문'].startswith('MOC')}
                          for x in o['그리드청산']],
        }
    })

payload = {
    'asOf': BW[-1]['d'].isoformat(),
    'rsi': BW[-1]['rsi'],
    'params': {'W_BASE': S.W_BASE, 'W_STRONG': S.W_STRONG, 'BO_K': S.BO_K,
               'BOS_K': S.BOS_K, 'BOS_RSI': S.BOS_RSI, 'BOS_MAX_W': S.BOS_MAX_W,
               'G1_DROP': S.G1_DROP, 'EPS': S.EPS, 'MAX_DAYS': S.MAX_DAYS,
               'TP_TOP': S.TP['TOP'], 'TP_BOTTOM': S.TP['BOTTOM'],
               'W_TOP': S.W_TOP, 'W_BOTTOM': S.W_BOTTOM},
    'cases': out,
}
p = os.path.join(H, '_parity_expected.json')
json.dump(payload, open(p, 'w', encoding='utf-8'), ensure_ascii=False, indent=1)
print(f'기대값 {len(out)}건 -> {p}')
print(f'기준일 {payload["asOf"]}  RSI {payload["rsi"]}')
print('이어서:  node verify/page_parity.js')
